### PR TITLE
Improve the definition and usage of the diff command

### PR DESCRIPTION
**Pull request by shindere**
This pull request is a follow-up to https://github.com/ocaml/ocaml/pull/10063#issuecomment-756665129

Currently, the build system defines a `DIFF` variable, but:

 * It is defined for the whole build system only on Windows platforms
 * On Unix it is deifned only in `Makefile.tools`
 * It contains both the name of the `diff` program and command-line options

In addition, it can be observed that:

 1. This `DIFF` variable is defined but not used any longer
 2. Although ocamltest does need to call a diff program to show differences between output and reference files, both the name of the program to invoke and the command-line options to use are hard-coded, so it does not used the value defined by the build system.

The present pull request fixes all this and is best reviewed commit per commit.

### CODE REVIEW

File: Makefile.config.in
Comment (dra27) [inline Makefile.config.in]: This should move away from the comment above it up to around L242 (below `NAKED_POINTERS`)

File: Makefile.config.in
Comment (dra27) [inline Makefile.config.in]: These were fully qualified in ce969ca16a4f2583fc27e7a5107841756e527618 apparently to fix AppVeyor. I expect this may have been something weird with another set of Cygwin utilities in `PATH` on AppVeyor.

**Review by dra27 — COMMENTED** [review]
Given that `DIFF` was inconsistent in `Makefile.config` before and these are only used by `ocamltest`, can `DIFF` and `DIFF_FLAGS` be put in `Makefile.build_config` (and so not installed)?

Two curiously connected thoughts:
- I would love ocamltest to use `patdiff` by default on my systems (I say by default, because I'll get to use it less if I have to remember to specify it!).
- `-u` is not a Posix option to `diff`!

These two things connected would make an `AC_CHECK_TOOL` for `patdiff, diff` be nice and the availability of `-u` can then be detected and added to the flags automatically (similarly, `--strip-trailing-cr` could then mandated by `configure`).

**Comment by shindere** [comment]
Many thanks for your review, David!

David Allsopp (2021/01/12 11:49 -0800):
> @dra27 commented on this pull request.
>
> Given that `DIFF` was inconsistent in `Makefile.config` before and
> these are only used by `ocamltest`, can `DIFF` and `DIFF_FLAGS` be put
> in `Makefile.build_config` (and so not installed)?

Sure, I can do that. Will post here when it's done (tomorrow).

> Two curiously connected thoughts:
> - I would love ocamltest to use `patdiff` by default on my systems (I
> say by default, because I'll get to use it less if I have to
> remember to specify it!).
> - `-u` is not a Posix option to `diff`!

Ah, din't know, thanks.

> These two things connected would make an `AC_CHECK_TOOL` for `patdiff,
> diff` be nice and the availability of `-u` can then be detected and
> added to the flags automatically (similarly, `--strip-trailing-cr`
> could then mandated by `configure`).

Yeah, I considered using configure but then thought it was perhaps a bit
overkill given that the tool is used by the testsuite only but I can do
that, sure.

> > @@ -260,11 +260,12 @@ endif # ifeq "$(TOOLCHAIN)" "msvc"
>  # User code should absolutely not rely on their presence because
>  # in the future their definition may be moved to a more private part of
>  # the compiler's build system
> +DIFF = diff
>
> This should move away from the comment above it up to around L242 (below `NAKED_POINTERS`)
>
> >  ifeq "$(UNIX_OR_WIN32)" "win32"
>    OTOPDIR=$(WINTOPDIR)
>    CTOPDIR=$(WINTOPDIR)
>    CYGPATH=cygpath -m
> -  DIFF=/usr/bin/diff -q --strip-trailing-cr
>
> These were fully qualified in ce969ca16a4f2583fc27e7a5107841756e527618
> apparently to fix AppVeyor. I expect this may have been something
> weird with another set of Cygwin utilities in `PATH` on AppVeyor.

Well. It was indeed fully qualified, but unused. The one hard-coded in
coamltest and effectively used was not fully qualified. So using a non
qualified name was intentionnal, to maintain ocamltest's behaviour. In
addition, I did try the qualified variant but `System.command` does not
like it, so if that's what we cant then I'd also need to change the way
the diff command is executed. It's possible of course but I tried to be
minimalistic at the moment.

Worth mentionning is that a working test suite will not test this,
because the command is invoked only when there is a difference between
reference file and output to show. I did however check manually that
things were still working correctly on a failing test both on a Windows
and a Unix platform.

**Comment by xavierleroy** [comment]
I just ran into the problem that this PR fixes: a test producing unexpect output in Alpine Linux, where ocamltest could not show the diff because Alpine's `diff` program doesn't have the `--strip-trailing-cr` option.

Could this PR be restarted and brought to a point where it can be merged?  Thanks.

**Comment by shindere** [comment]
Sure, I'll happily restart working on it not later than tomorrow.

**Comment by shindere** [comment]
Okay, it took a bit longer than expected, sorry about that, but here is
an updated implementation.

Since the diff tool is used only in ocamltest, its detection takes place
in configure only if ocamltest is enabled.

configure first looks for patdiff and, it it is not found, for diff. If
none of those is found that's a configure error.

If it's diff which is found, the `-u` and `--strip-trailing-cr` options
are tested. If they work they are recorded in the build system so that
ocamltest can pass them to diff when it invokes it.

In other words, a diff implementation not supporting `-u` will not be
called with -u, with this PR.

**Comment by shindere** [comment]
Just pushed an enhanced version: configure prints one message per
diff option which is tested and the result of the test (whether the option
is found to be supported or not).

File: configure.ac
Comment (dra27) [inline configure.ac]: Oh dear, this message (written by me in #9250) was pretty dreadful. I think adding the word test makes it more confusing - could it more simply just be "do not build ocamltest"?

File: configure.ac
Comment (dra27) [inline configure.ac]: Up to you, but I think this is slightly less brittle?
```suggestion
AS_IF([test -n "$1" ],
```

File: configure.ac
Comment (dra27) [inline configure.ac]: Please could this be done with `AS_IF` - I found the shell `if` with a m4sh `[` initially confusing.

**Review by dra27 — COMMENTED** [review]
Looks good to me - I have a few minor suggestions, but nothing fundamental.

**Comment by shindere** [comment]
David Allsopp (2022/03/14 06:30 -0700):
> @dra27 commented on this pull request.
>
> Looks good to me - I have a few minor suggestions, but nothing fundamental.
>
> > @@ -302,7 +303,7 @@ AC_ARG_WITH([odoc],
>
>  AC_ARG_ENABLE([ocamltest],
>    [AS_HELP_STRING([--disable-ocamltest],
> -    [do not build the ocamltest driver])])
> +    [do not build the ocamltest test driver])])
>
> Oh dear, this message (written by me in #9250) was pretty dreadful. I
> think adding the word test makes it more confusing - could it more
> simply just be "do not build ocamltest"?

Done so.

In the same vein, there is also "do not build the ocamldoc documentation
system", to some extent "do not build the native compiler" (we could
either write ocamlopt, or the native compilerS), "do not build or
install the library man pages" which I find a bit unclear (if it means
not build then of course it won't be installed, if it means not install
then we could just write not install), and perhaps also "build the
debugger" which we could write "build ocamldebug".

> > @@ -1878,6 +1879,24 @@ AS_CASE([$enable_ocamltest,OCAML__DEVELOPMENT_VERSION],
>    [yes,*|,true],[ocamltest='ocamltest'],
>    [ocamltest=''])
>
> +AS_IF([test x"$ocamltest" = "xocamltest"],
>
> Up to you, but I think this is slightly less brittle?
> ```suggestion
> AS_IF([test -n "$1" ],
> ```

I took the `-n` idea into account but not the `$1` because keeping the
explicit reference to `$ocamltest` seems more readable to me and moe
robust, too, in case the test is moved or something is inserted between
it and the block just before.

>
> > @@ -1878,6 +1879,24 @@ AS_CASE([$enable_ocamltest,OCAML__DEVELOPMENT_VERSION],
>    [yes,*|,true],[ocamltest='ocamltest'],
>    [ocamltest=''])
>
> +AS_IF([test x"$ocamltest" = "xocamltest"],
> +  [AC_CHECK_PROGS([DIFF], [patdiff diff], [none])
> +  DIFF_FLAGS=""
> +  flags=""
> +  AS_CASE([$DIFF],
> +    [diff], [flags="--strip-trailing-cr -u"],
> +    [none],
> +      [AC_MSG_ERROR([ocamltest requires a diff tool])])
> +  for flag in ${flags}; do
> +    AC_MSG_CHECKING([whether $DIFF supports $flag]);
> +    if [$DIFF $DIFF_FLAGS $flag $0 $0 > /dev/null 2>&1]; then
>
> Please could this be done with `AS_IF` - I found the shell `if` with a
> m4sh `[` initially confusing.

Sure, done.

I pushed the changes as three separate commits to (hopefully) ease
reviewing. Once everything is okay I'll squash them into the first
commit of the initial PR before merging, so please do not merge the PR
in its current state, I'll do so once it got approved.

**Review by dra27 — APPROVED** [review]
LGTM, thanks!

> In the same vein, there is also "do not build the ocamldoc documentation
system", to some extent "do not build the native compiler" (we could
either write ocamlopt, or the native compilerS), "do not build or
install the library man pages" which I find a bit unclear (if it means
not build then of course it won't be installed, if it means not install
then we could just write not install), and perhaps also "build the
debugger" which we could write "build ocamldebug".

I briefly thought about that earlier - `ocamldoc` is more than just the `ocamldoc` binary (there's libraries and plugins and so forth installed too) so its description as "the ocamldoc documentation system" seems OK. The other two could be changed, although neither reads as confusingly as what I'd written previously for ocamltest!

> but not the `$1` because keeping the
explicit reference to `$ocamltest` seems more readable

I don't know what misfired in my brain to type the `$1` in the suggestion! `$ocamltest` both more readable and more correct.

**Pull request merged**
